### PR TITLE
Include location state in LOCATION_CHANGE payload

### LINKED ISSUE #301: Location state absent in LOCATION_CHANGE action
React-router uses history's `location` object:

>The `location` object implements a subset of [the `window.location` interface](https://developer.mozilla.org/en-US/docs/Web/API/Location), including:
>
>- `location.pathname` - The path of the URL
>- `location.search` - The URL query string
>- `location.hash` - The URL hash fragment
>
>Locations may also have the following properties:
>
>- `location.state` - Some extra state for this location that does not reside in the URL (supported in `createBrowserHistory` and `createMemoryHistory`)
>- `location.key` - A unique string representing this location (supported in `createBrowserHistory` and `createMemoryHistory`)

(see https://github.com/ReactTraining/history#listening)

connected-react-router only passes `pathname`, `search` and `hash` fields to the `LOCATION_CHANGE` action, and ignores the `state` and `key` fields.

https://github.com/supasate/connected-react-router/blob/37e90d7a55717046d4265f5832d3f9c8356e7db8/src/ConnectedRouter.js#L28-L49

As a consequence, a link using location state like this:

```jsx
<Link to={{ pathname: '/hello', state: { foo: 'bar' } }}>Hello</Link>
```

Is translated by connected-react-router as the following action:

```js
{
  type: '@@router/LOCATION_CHANGE',
  payload: {
    location: {
      pathname: '/hello',
      search: '',
      hash: ''
    },
    action: 'POP',
    isFirstRendering: false
  }
}
```

Notice the missing `state` key in the `location`.

As a consequence, reducers can't use the value passed in the `state`.

Note that [this was properly handled by `react-router-redux`](https://github.com/ReactTraining/react-router/blob/v4.3.1/packages/react-router-redux/modules/ConnectedRouter.js).

### PR DESCRIPTION
**Pull request by fzaninotto**
- [x] Add code
- [x] Add test
- [x] Run all tests to check for regression

Closes #301

### CODE REVIEW

**Comment by fzaninotto** [comment]
Is there anything I can do to help this PR to be merged?

**Comment by fzaninotto** [comment]
bump, can you please merge this PR?

**Comment by rob2d** [comment]
would also like the sanity introduced with this feature 🙏 at the moment you need to hard code your param matching or find an extremely over-engineered/coupled way to do that with view.

**Comment by akshah123** [comment]
@supasate any chance this can be merged in?

File: test/ConnectedRouter.test.js [hunk 1]
```diff
@@ -123,6 +123,19 @@ describe('ConnectedRouter', () => {
...
 
+    it('supports location state and key', () => {
+      mount(
+        <Provider store={store}>
+          <ConnectedRouter {...props}>
+            <Route path="/" render={() => <div>Home</div>} />
+          </ConnectedRouter>
+        </Provider>
+      )
+      props.history.push({ pathname: '/new-location', state: { foo: 'bar' } })
```
Comment (supasate) [inline test/ConnectedRouter.test.js]: Can you test `key` as well?

**Comment by supasate** [comment]
Hi @fzaninotto, sorry for keeping this long. I somehow thought it was a stale PR and forgot to re-check. 

LGTM with some nits for the unit test. Please rebase as well.

File: test/ConnectedRouter.test.js (hunk 1 above)
Comment (fzaninotto) [reply test/ConnectedRouter.test.js]: I removed `location.key` from the perimeter of this PR, as it seems to be a private API of the `history` package.

**Comment by fzaninotto** [comment]
Rebased

**Comment by supasate** [comment]
LGTM. Thanks for your contribution!

**Comment by fzaninotto** [comment]
Awesome! Any ETA on the release of this fix?

**Comment by akshah123** [comment]
@supasate would it be possible release this fix soon?

**Comment by supasate** [comment]
Released in 6.7.0.

**Pull request merged**